[New Tutorial] EKS tutorial Monitor Container Applications (#731)
build-on-aws/content: tutorials/eks-monitor-containerized-applications/index.md

---
title: "Easily Monitor Containerized Applications with Amazon CloudWatch Container Insights"
description: "How to collect, aggregate, and analyze metrics from your containerized applications using Amazon CloudWatch Container Insights."
tags:
    - eks-container-insight
    - eks
    - kubernetes
    - tutorials
    - aws
showInHomeFeed: true
waves:
  - modern-apps
spaces:
  - kubernetes
  - modern-apps
authorGithubAlias: berry2012
authorName: Olawale Olaleye
additionalAuthors: 
  - authorGithubAlias: ahmadt312
    authorName: Ahmad Tariq
date: 2023-09-30
---

> Monitoring containerized applications requires precision and efficiency. Amazon CloudWatch Container Insights is an essential tool that enables this monitoring, handling the complexities of collecting and summarizing metrics from your applications. These metrics are especially critical to the smooth operation of containerized microservices running on Kubernetes, like with Amazon EKS. Critical metrics you need to collect and monitor include utilization for resources such as CPU, memory, disk, and network. Container Insights also provides diagnostic information, such as container restart failures, to help you isolate issues and resolve them quickly. This tutorial focuses on the integration of CloudWatch Container Insights within an Amazon EKS cluster, deploying container applications, and monitoring the performance of the application with Container Insights. Picture a complex system that needs to manage multiple metrics, or a microservices architecture that requires meticulous performance tracking. With Amazon CloudWatch Container Insights, the complexities of managing container metrics are drastically reduced. This fully integrated monitoring solution allows you the time to focus on enhancing your application's core functionalities. As the performance metrics of your containers change, Container Insights offers real-time data, enabling you to maintain consistent application performance through informed decisions.

Reviewer: This first paragraph is a bit repetitive and feels a little sales-y. Consider something like this to get to the important part more quickly:

Monitoring containerized applications requires precision and efficiency. One way to handle the complexities of collecting and summarizing metrics from your applications is to use Amazon CloudWatch Container Insights. As the performance metrics of your containers change, Container Insights offers real-time data, enabling you to maintain consistent application performance through informed decisions.

> Building on the Amazon EKS cluster from part 1 of our series, this tutorial dives into setting up Amazon CloudWatch Container Insights. Included in the cluster configuration for the previous tutorial is the [Amazon CloudWatch](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/WhatIsCloudWatch.html?sc_channel=el&sc_campaign=appswave&sc_content=eks-cluster-high-traffic&sc_geo=mult&sc_country=mult&sc_outcome=acq) IAM policy attached to the IAM Role for Service Account (IRSA) and the [OpenID Connect (OIDC) endpoint](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html?sc_channel=el&sc_campaign=appswave&sc_content=eks-cluster-high-traffic&sc_geo=mult&sc_country=mult&sc_outcome=acq). For part one of this series, see [Building an Amazon EKS Cluster Preconfigured to Run High Traffic Microservices](https://community.aws/tutorials/eks-cluster-high-traffic). Alternatively, to setup an existing cluster with the components required for this tutorial, use the instructions in [Verify prerequisites](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-prerequisites.html) in EKS official documentation.

Reviewer: ```suggestion
Building on the Amazon EKS cluster from part 1 of our series, this tutorial dives into setting up Amazon CloudWatch Container Insights. Included in the cluster configuration for the previous tutorial is the [Amazon CloudWatch](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/WhatIsCloudWatch.html?sc_channel=el&sc_campaign=appswave&sc_content=eks-cluster-high-traffic&sc_geo=mult&sc_country=mult&sc_outcome=acq) IAM policy attached to the IAM Role for Service Account (IRSA) and the [OpenID Connect (OIDC) endpoint](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html?sc_channel=el&sc_campaign=appswave&sc_content=eks-cluster-high-traffic&sc_geo=mult&sc_country=mult&sc_outcome=acq). For part one of this series, see [Building an Amazon EKS Cluster Preconfigured to Run High Traffic Microservices](https://community.aws/tutorials/eks-cluster-high-traffic). Alternatively, to set up an existing cluster with the components required for this tutorial, use the instructions in [the _verify prerequisites_ section](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-prerequisites.html) of EKS official documentation.
```

| Attributes             |                                                                 |
|------------------------|-----------------------------------------------------------------|
| ✅ AWS experience      | 200 - Intermediate                                              |
| ⏱ Time to complete     | 30 minutes                                                      |
| 🧩 Prerequisites       | - [AWS Account](https://aws.amazon.com/resources/create-account/?sc_channel=el&sc_campaign=appswave&sc_content=eks-cluster-high-traffic&sc_geo=mult&sc_country=mult&sc_outcome=acq)|
| 📢 Feedback            | <a href="https://www.pulse.aws/survey/Z8XBGQEL" target="_blank">Any feedback, issues, or just a</a> 👍 / 👎 ?    |
| ⏰ Last Updated        | 2023-09-30                                                      |

| ToC |
|-----|

## Prerequisites

Before you begin this tutorial, you need to:

* Install the latest version of [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl). To check your version, run: `kubectl version --short`.
* Install the latest version of [eksctl](https://eksctl.io/introduction/#installation). To check your version, run: `eksctl info`.

> ## Overview

Reviewer: ```suggestion
```

> The tutorial focuses on configuring Container Insights on Amazon EKS, deploying a container application in the cluster, and monitoring the application's performance using Container Insights.

Reviewer: ```suggestion
```

## Step 1: Set up Container Insights on Amazon EKS

For CloudWatch Container Insights to collect, aggregate, and summarize metrics and logs from your containerized applications and microservices on Amazon Elastic Kubernetes Service (Amazon EKS), some setup steps need to be performed. Container Insights supports both Amazon EKS EC2 and [Fargate](https://docs.aws.amazon.com/eks/latest/userguide/fargate-logging.html). There are a few ways you can set up Container Insights on an Amazon EKS cluster: using the CloudWatch agent, a “quick start” setup, or through a manual setup approach. Below, you will find the steps required for the “quick start” method.

> ### Quick Start Set Up

Reviewer: ```suggestion
### Quick Start Setup
```

NAME                                          READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/quickstart-nginx-deployment   3/3     3            3           17s

NAME                                                     DESIRED   CURRENT   READY   AGE
replicaset.apps/quickstart-nginx-deployment-7cd757dc7b   3         3         3       17s
```

4. Use the following command to view the real-time logs of the "load" Pod, which is continuously making requests to the Nginx service. Use Ctrl+C to stop.

```bash
kubectl logs -f load -n quickstart
```

The expected output should look like this:

```bash
PodName: quickstart-nginx-deployment-7cd757dc7b-wpw4x NodeName: ip-192-168-141-57.us-east-2.compute.internal podIP: 192.168.136.230
PodName: quickstart-nginx-deployment-7cd757dc7b-fv592 NodeName: ip-192-168-177-109.us-east-2.compute.internal podIP: 192.168.164.31
PodName: quickstart-nginx-deployment-7cd757dc7b-fv592 NodeName: ip-192-168-177-109.us-east-2.compute.internal podIP: 192.168.164.31
PodName: quickstart-nginx-deployment-7cd757dc7b-9fss6 NodeName: ip-192-168-119-7.us-east-2.compute.internal podIP: 192.168.112.25
```

> 

Reviewer: ```suggestion
```

## Step 3: Use [CloudWatch Logs Insights](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/AnalyzingLogData.html) Query to search and analyze container logs

> You can use [CloudWatch Logs Insights](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/AnalyzingLogData.html) Query to interactively search and analyze the container logs of the application in Amazon CloudWatch Logs. Fluent Bit send logs from your containers in the cluster to CloudWatch Logs. In the Step 1 above, we’ve set up Fluent Bit as a DaemonSet to send logs to CloudWatch Logs. Fluent Bit creates the log group below if it don't already exist:

Reviewer: ```suggestion
You can use [CloudWatch Logs Insights](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/AnalyzingLogData.html) Query to interactively search and analyze the container logs of the application in Amazon CloudWatch Logs. Fluent Bit sends logs from your containers in the cluster to CloudWatch Logs. In Step 1 above, we’ve set up Fluent Bit as a DaemonSet to send logs to CloudWatch Logs. Fluent Bit creates the log group below if it doesn't already exist:
```

`/aws/containerinsights/Cluster_Name/application` which contains all log files in `/var/log/containers` on each worker node in the cluster.

> ### To run a CloudWatch Logs Insights sample query:

Reviewer: ```suggestion
### To Run a CloudWatch Logs Insights Sample Query:
```

* Open the CloudWatch console.
* In the navigation pane, choose **Logs**, and then choose **Log groups**.
* Click the log group `/aws/containerinsights/CLUSTER_NAME/application`. Where CLUSTER_NAME is the actual name of your EKS cluster.

> * Under the log details (top-right), click **View in Logs Insights** 

Reviewer: ```suggestion
* Under the log details (top-right), click **View in Logs Insights**.
```

* Delete the default query in the CloudWatch Log Insight Query Editor. Then, enter the following command and select Run query:

```bash
fields @timestamp, kubernetes.pod_name as PodName, kubernetes.host as WorkerNode, kubernetes.namespace_name as Namespace, log
| filter PodName like 'quickstart-nginx-deployment'
| sort @timestamp desc
| limit 200
```

> * Use the time interval selector to select a time period that you want to query. For example:

Reviewer: ```suggestion

* Use the time interval selector to select a time period that you want to query. For example:
```

> ![](./images/Container_Insight_Query.png)

Reviewer: Please add alt text between the brackets to describe what appears in the image, as though the reader is unable to see the image.

> 

Reviewer: ```suggestion
```

> ## Step 4: Monitor performance of the application with Container Insights

Reviewer: ```suggestion
## Step 4: Monitor Performance of the Application with Container Insights
```

### View Container Insights Dashboard Metrics

In this section, you will learn how to access the Container Insights Dashboard Metrics within AWS CloudWatch. This dashboard provides a centralized view of your Amazon EKS and Kubernetes clusters' performance, offering real-time insights into various metrics such as CPU utilization, memory usage, and network activity. By following these steps, you can quickly navigate to the specific cluster and resources you wish to monitor, enabling you to keep a close eye on the health and performance of your containerized applications.

1. Open the CloudWatch console at https://console.aws.amazon.com/cloudwatch/.
2. In the left navigation pane, open the **Insights** dropdown menu, and then choose **Container Insights**.
3. Under “Container Insights” (top), select **Performance Monitoring** from the dropdown menu.
4. In the “EKS Clusters” dropdown field, select the name of your cluster.
5. Use the additional dropdown menus to filter resources, such as “EKS Clusters” and “EKS Pods.” For example:

> ![](./images/Container_Insights.png)

Reviewer: Please add alt text between the brackets to describe what appears in the image, as though the reader is unable to see the image.

> ### View additional[ Amazon EKS and Kubernetes Container Insights metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-metrics-EKS.html)

Reviewer: ```suggestion
### View Additional [Amazon EKS and Kubernetes Container Insights Metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-metrics-EKS.html)
```

> In this section, you will explore how to access a broader set of metrics specific to Amazon EKS and Kubernetes within AWS CloudWatch. These additional metrics like pod_cpu_utilization_over_pod_limit, provide deeper insights into the performance and behavior of your containerized applications and the underlying infrastructure. Whether you are looking to analyze CPU utilization, Memory usage, or Network metrics, this process allows you to customize your view and focus on the aspects most relevant to your needs.

Reviewer: ```suggestion
In this section, you will explore how to access a broader set of metrics specific to Amazon EKS and Kubernetes within AWS CloudWatch. These additional metrics, like pod_cpu_utilization_over_pod_limit, provide deeper insights into the performance and behavior of your containerized applications and the underlying infrastructure. Whether you are looking to analyze CPU utilization, Memory usage, or Network metrics, this process allows you to customize your view and focus on the aspects most relevant to your needs.
```

> 

Reviewer: ```suggestion
```

> #### Let’s explore an application exceeding its resource limit

Reviewer: ```suggestion
#### Let’s Explore an Application Exceeding Its Resource Limit
```

```bash
NAME                           READY   STATUS              RESTARTS       AGE
geo-api-load-c9c7bf98c-4rrn8   0/1     ContainerCreating   0               0s
geo-api-load-c9c7bf98c-kzsjs   0/1     ContainerCreating   0               0s
geo-api-load-c9c7bf98c-4rrn8   1/1     Running             0               1s
geo-api-load-c9c7bf98c-kzsjs   1/1     Running             0               2s
geo-api-76f6dcf999-ptpz5       1/1     Running             20 (5m8s ago)   118m
geo-api-76f6dcf999-ptpz5       0/1     OOMKilled           20 (5m13s ago)   118m
```

The output shows that the geo-api Pod status is Running and sometimes getting OOMKilled.

Get a more detailed view of the Container status:

```bash
kubectl get pod geo-api-76f6dcf999-ptpz5 --output=yaml | grep -i lastState -A7
```

The output shows that the Container was killed because it is out of memory (OOM):

```bash
    lastState:
      terminated:
        containerID: containerd://4bbdfee06a3d3daca0e74f14f18f8a66ac0a415c79720eae44ea9ad4c46bcb37
        exitCode: 137
        finishedAt: "2023-08-26T12:48:37Z"
        reason: OOMKilled
        startedAt: "2023-08-26T12:47:27Z"
    name: geo-api
```

> 5. Let’s view the Container Insights metrics of this pod

Reviewer: ```suggestion
5. Let’s view the Container Insights metrics of this pod:
```

    1. Open the CloudWatch console at https://console.aws.amazon.com/cloudwatch/.
    2. In the navigation pane, choose **Metrics**, and then choose **All metrics**.

>     3. Select the **ContainerInsights** metric namespace. Select the **ClusterName**, **Namespace**, and **PodName**, in the search bar, copy and paste **PodName="geo-api**" 

Reviewer: ```suggestion
    3. Select the **ContainerInsights** metric namespace. Select the **ClusterName**, **Namespace**, and **PodName**, in the search bar, copy and paste **PodName="geo-api**.
```

    4. You can view the percentage of CPU units being used by the pod relative to the pod limit and the percentage of memory that is being used by pods relative to the pod limit by selecting the metrics below:
        * `pod_cpu_utilization_over_pod_limit`
        * `pod_memory_utilization_over_pod_limit`

> ![](./images/resource_monitoring.png)

Reviewer: Needs alt text

The graph shows that the container in the pod has completely utilized its CPU and memory limit and you will need to specify enough resources to prevent the container in the pod from being terminated.

> 

Reviewer: ```suggestion
```

## Clean Up

> To avoid incurring future charges, you should delete the resources created during this tutorial. You can delete the container application and the also delete the CloudWatch agent and Fluent bit for Container Insights with the following command. 

Reviewer: ```suggestion
To avoid incurring future charges, you should delete the resources created during this tutorial. You can delete the container application and also delete the CloudWatch agent and Fluent bit for Container Insights with the following command:
```

```bash
# Delete workloads
kubectl delete -f workload.yaml -n quickstart

kubectl delete -f geo-api.yaml

# Delete the the CloudWatch agent and Fluentbit for Container Insights
kubectl delete -f cwagent-fluent-bit-quickstart.yaml
```

## Conclusion

> By following this tutorial, you've successfully set up CloudWatch agent and Fluent Bit for Container Insights to monitor  sample containerized workloads in an Amazon EKS cluster. With these instructions, you'll have a robust monitoring and logging solution to help you monitor the performance of application deployments in the cluster.

Reviewer: Please provide one or two links for further reading if the viewer wants to explore more on the topic!
